### PR TITLE
[feature](Nereids)add type coercion rule for nereids

### PR DESCRIPTION
**Pull request by morrySnow**
# Proposed changes

- add an interface ExpectsInputTypes to Expression
- add an interface ImplicitCastInputTypes to Expression
- add a Expression rewrite rule for type coercion
- add a Check Analysis Rule to check whether Plan is Semantically correct

if Expression implements ImplicitCastInputTypes, type coercion rule will automatic rewrite its children that casting it to the most suitable type.
If Expression implements ExpectsInputTypes, Check Analysis will check its children's type whether accepted by expects input types.

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
2. Has unit tests been added:
    - [x] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [x] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/InSubquery.java [hunk 1]
```diff
@@ -70,7 +71,7 @@ public Expression getCompareExpr() {
     @Override
     public Expression withChildren(List<Expression> children) {
         Preconditions.checkArgument(children.size() == 2);
-        Preconditions.checkArgument(children.get(0) instanceof Expression);
+        Preconditions.checkArgument(children.get(0) != null);
```
Comment (Kikyou1997) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/InSubquery.java:74]: I think this uncheck is unneccessary, it's already checked in the constructor of InSubQuery

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/BinaryOperator.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Abstract for all binary operator, include binary arithmetic, compound predicate, comparison predicate.
+ */
+public abstract class BinaryOperator extends Expression implements BinaryExpression, ExpectsInputTypes {
```
Comment (Kikyou1997) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/BinaryOperator.java:32]: Rename to AbstractBinaryExpression?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/Avg.java [hunk 3]
```diff
@@ -18,25 +18,46 @@
...
     public Avg(Expression child) {
         super("avg", child);
     }
 
     @Override
     public DataType getDataType() {
-        return DoubleType.INSTANCE;
+        if (child().getDataType() instanceof DecimalType) {
+            DecimalType childType = (DecimalType) child().getDataType();
+            return DecimalType.createDecimalType(childType.getPrecision(), childType.getScale());
```
Comment (Kikyou1997) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/Avg.java:53]: why not return child type directly

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/Substring.java [hunk 4]
```diff
@@ -18,19 +18,30 @@
...
+import org.apache.doris.nereids.types.coercion.AbstractDataType;
+import org.apache.doris.nereids.types.coercion.TypeCollection;
 
 import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
 
 import java.util.List;
 
 /**
  * substring function.
```
Comment (Kikyou1997) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/Substring.java:36]: Add some examples

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/Substring.java [hunk 5]
```diff
@@ -18,19 +18,30 @@
...
  */
-public class Substring extends BoundFunction implements TernaryExpression {
+public class Substring extends BoundFunction implements TernaryExpression, ImplicitCastInputTypes {
+
+    private static final List<AbstractDataType> EXPECTED_INPUT_TYPES = ImmutableList.of(
```
Comment (Kikyou1997) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/Substring.java:40]: comment pls

**Comment by morrySnow** [comment]
@Kikyou1997 PTAL

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/InSubquery.java (hunk 1 above)
Comment (morrySnow) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/InSubquery.java:74]: do not check in constructor

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/BinaryOperator.java (hunk 2 above)
Comment (morrySnow) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/BinaryOperator.java:32]: it is not abstract binary expression at all. such as InSubquery is not a subclass of this class

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/Substring.java (hunk 4 above)
Comment (morrySnow) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/Substring.java:36]: all function's comment should update. So let's do it later

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/TypeCoercion.java [hunk 6]
```diff
@@ -0,0 +1,394 @@
...
+
+/**
+ * a rule to add implicit cast for expressions.
+ */
+@Developing
+public class TypeCoercion extends AbstractExpressionRewriteRule {
+
+    public static final TypeCoercion INSTANCE = new TypeCoercion();
+
+    public static final List<DataType> NUMERIC_PRECEDENCE = ImmutableList.of(
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/TypeCoercion.java:70]: Add comment to explain the precedence order.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/TypeCoercion.java [hunk 7]
```diff
@@ -0,0 +1,394 @@
...
+        } else if (input.isPrimitive() && !input.isStringType() && expected instanceof CharacterType) {
+            returnType = StringType.INSTANCE;
+        }
+
+        // could not do implicit cast, just return null. Throw exception in check analysis.
+        return Optional.ofNullable(returnType);
+    }
+
+    private boolean canHandleTypeCoercion(DataType leftType, DataType rightType) {
+        if (leftType instanceof DecimalType && rightType instanceof NullType) {
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/TypeCoercion.java:260]: Why can not do type convert for decimal on both side?
Add comment to explain.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/TypeCoercion.java [hunk 8]
```diff
@@ -0,0 +1,394 @@
...
+        if (!(leftType instanceof DecimalType) && !(rightType instanceof DecimalType) && !leftType.equals(rightType)) {
+            return true;
+        }
+        return false;
+    }
+
+    // TODO: compatible with origin planner and BE
+    // TODO: when add new type, add it to here
+    @Developing
+    private Optional<DataType> findTightestCommonType(DataType left, DataType right) {
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/TypeCoercion.java:275]: Missing Date/Datetime?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/TypeCoercion.java [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+            return binaryOperator.withChildren(left, right);
+        }
+        return TypeCoercionUtils.findTightestCommonType(left.getDataType(), right.getDataType())
+                .map(commonType -> {
+                    if (binaryOperator.inputType().acceptsType(commonType) && (
+                            !left.getDataType().equals(commonType) || !right.getDataType().equals(commonType))) {
+                        Expression newLeft = left;
+                        Expression newRight = right;
+                        if (!left.getDataType().equals(commonType)) {
+                            newLeft = new Cast(left, commonType);
```
Comment (englefly) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/TypeCoercion.java:73]: from efficiency perspective, here we will traverse the left child top-down again. Shall we combine this traverse with the traverse in line 61?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/util/TypeCoercionUtils.java [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+     */
+    @Developing
+    public static Optional<DataType> findTightestCommonType(DataType left, DataType right) {
+        // TODO: compatible with origin planner and BE
+        // TODO: when add new type, add it to here
+        DataType tightestCommonType = null;
+        if (left.equals(right)) {
+            tightestCommonType = left;
+        } else if (left instanceof NullType) {
+            tightestCommonType = right;
```
Comment (englefly) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/util/TypeCoercionUtils.java:155]: when left and right are both nullable, why we choose right?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/util/TypeCoercionUtils.java [hunk 11]
```diff
@@ -0,0 +1,292 @@
...
+        } else if (right instanceof NullType) {
+            tightestCommonType = left;
+        } else if (left instanceof IntegralType && right instanceof DecimalType
+                && ((DecimalType) right).isWiderThan(left)) {
```
Comment (englefly) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/util/TypeCoercionUtils.java:159]: do we missed the case:
left is integral and right is decimal but  (NOT right.isWiderThan(left))

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/TypeCoercion.java (hunk 8 above)
Comment (morrySnow) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/TypeCoercion.java]: yes, will add it in next PR about type coercion

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/TypeCoercion.java (hunk 7 above)
Comment (morrySnow) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/TypeCoercion.java]: add in next PR

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/TypeCoercion.java (hunk 9 above)
Comment (morrySnow) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/TypeCoercion.java:73]: we could add a memorize supplier to do that just like what plan's output do. I think we could optimize it in a individual PR

File: fe/fe-core/src/main/java/org/apache/doris/nereids/util/TypeCoercionUtils.java (hunk 10 above)
Comment (morrySnow) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/util/TypeCoercionUtils.java:155]: this is not nullable, but is NullType. NullType is just for NullLiteral

File: fe/fe-core/src/main/java/org/apache/doris/nereids/util/TypeCoercionUtils.java (hunk 11 above)
Comment (morrySnow) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/util/TypeCoercionUtils.java:159]: it will be processed by findWiderTypeForDecimal

**Review by 924060929 — COMMENTED** [review]
you can uncomment the `order_qt_in_predicate_5` and `order_qt_in_predicate_6` in `regression-test/suites/nereids_syntax_p0/inpredicate.groovy` for regression testing

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/ComparisonPredicate.java [hunk 12]
```diff
@@ -39,45 +37,21 @@ public abstract class ComparisonPredicate extends Expression implements BinaryEx
...
-            return true;
-        }
-        if (o == null || getClass() != o.getClass()) {
-            return false;
-        }
-        ComparisonPredicate other = (ComparisonPredicate) o;
-        return Objects.equals(left(), other.left())
-                && Objects.equals(right(), other.right());
+    public AbstractDataType inputType() {
+        return AnyDataType.INSTANCE;
```
[the remark this answers is not in the document]
Comment (morrySnow) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/ComparisonPredicate.java:54]: this is comparison, such as equalTo, so need BooleanType.INSTANCE

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/CheckAnalysis.java [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+    private Plan checkExpressionInputTypes(Plan plan) {
+        final Optional<TypeCheckResult> firstFailed = plan.getExpressions().stream()
+                .map(Expression::checkInputDataTypes)
+                .filter(TypeCheckResult::failed)
+                .findFirst();
+
+        if (firstFailed.isPresent()) {
+            throw new RuntimeException(firstFailed.get().getMessage());
```
[the remark this answers is not in the document]
Comment (morrySnow) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/CheckAnalysis.java:44]: ok, btw exception will be refactor totally after we could generate best state-of-the-art plan on TPC-H benchmark

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/TypeCoercion.java [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+     */
+    @Developing
+    private Optional<Expression> implicitCast(Expression input, AbstractDataType expected,
+            ExpressionRewriteContext ctx) {
+        Expression rewrittenInput = rewrite(input, ctx);
+        Optional<DataType> castDataType = TypeCoercionUtils.implicitCast(rewrittenInput.getDataType(), expected);
+        if (castDataType.isPresent() && !castDataType.get().equals(rewrittenInput.getDataType())) {
+            return Optional.of(new Cast(rewrittenInput, castDataType.get()));
+        } else {
+            // TODO: there maybe has performance problem, we need use ctx to save whether children is changed.
```
[the remark this answers is not in the document]
Comment (morrySnow) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/TypeCoercion.java:172]: good idea

**Comment by morrySnow** [comment]
> inpredicate.groovy

done

**Comment by morrySnow** [comment]
> you can uncomment the `order_qt_in_predicate_5` and `order_qt_in_predicate_6` in `regression-test/suites/nereids_syntax_p0/inpredicate.groovy` for regression testing

done

**Review by 924060929 — APPROVED** [review]
LGTM

**Pull request merged**